File: Live Class - Java/Producer_Consumer_Solution.java
import java.util.LinkedList;
import java.util.Random;

class Buffer {
    private final LinkedList<Integer> buffer = new LinkedList<>();
    private final int bufferSize;

    public Buffer(int size) {
        this.bufferSize = size;
    }

    public void produce(int item) throws InterruptedException {
        synchronized (this) {
            while (buffer.size() == bufferSize) {
                wait(); // Wait until there is space in the buffer
            }
            buffer.add(item);
            System.out.println("Produced: " + item);
            notify(); // Notify a consumer that an item has been produced
        }
    }

    public int consume() throws InterruptedException {
        synchronized (this) {
            while (buffer.isEmpty()) {
                wait(); // Wait until there is an item to consume
            }
            int item = buffer.removeFirst();
            System.out.println("Consumed: " + item);
            notify(); // Notify a producer that there is space in the buffer
            return item;
        }
    }
}

class Producer extends Thread {
    private final Buffer buffer;

    public Producer(Buffer buffer) {
        this.buffer = buffer;
    }

    @Override
    public void run() {
        Random random = new Random();
        try {
            while (true) {
                int item = random.nextInt(100) + 1; // Produce a random item
                buffer.produce(item);
                Thread.sleep((long) (random.nextDouble() * 900) + 100); // Sleep between 0.1 to 1 second
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}

class Consumer extends Thread {
    private final Buffer buffer;

    public Consumer(Buffer buffer) {
        this.buffer = buffer;
    }

    @Override
    public void run() {
        try {
            while (true) {
                buffer.consume();
                Thread.sleep((long) (Math.random() * 900) + 100); // Sleep between 0.1 to 1 second
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}

public class Producer_Consumer_Solution {
    public static void main(String[] args) {
        Buffer buffer = new Buffer(10);
        Producer producer = new Producer(buffer);
        Consumer consumer = new Consumer(buffer);

        producer.start();
        consumer.start();
    }
}

File: Live Class - Python/producer_consumer_solution.py
#Producer Consumer Solution : Slide 50
import threading
import time
import random

# Shared resource
buffer = []
buffer_size = 10

# Condition object for synchronization
condition = threading.Condition()

class Producer(threading.Thread):
    def run(self):
        global buffer
        while True:
            item = random.randint(1, 100)
            condition.acquire()
            while len(buffer) == buffer_size:
                condition.wait()
            buffer.append(item)
            print(f"Produced: {item}")
            condition.notify()
            condition.release()
            time.sleep(random.uniform(0.1, 1))
class Consumer(threading.Thread):
    def run(self):
        global buffer
        while True:
            condition.acquire()
            while not buffer:
                condition.wait()
            item = buffer.pop(0)
            print(f"Consumed: {item}")
            condition.notify()
            condition.release()
            time.sleep(random.uniform(0.1, 1))
if __name__ == "__main__":
    producer = Producer()
    consumer = Consumer()
    
    producer.start()
    consumer.start()
    
    producer.join()
    consumer.join()


